### PR TITLE
 Support various BERT relative position embeddings

### PR DESCRIPTION
**Pull request by zhiheng-huang**
# What does this PR do?

The default BERT model `bert-base-uncased` was pre-trained with absolute position embeddings. We provide three pre-trained models which were pre-trained on the same training data (BooksCorpus and English Wikipedia) as in the BERT model training, but with different relative position embeddings (Shaw et al., Self-Attention with Relative Position Representations, https://arxiv.org/abs/1803.02155 and Huang et al., Improve Transformer Models with Better Relative Position Embeddings, https://arxiv.org/abs/2009.13658, accepted in findings of EMNLP 2020). We show how to fine-tune these pre-trained models on SQuAD1.1 data set and we also report the EM and F1 score on SQuAD1.1 dev dataset. See examples/question-answering/README.md for more details.

Fixes # (issue)

N/A

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@LysandreJik @julien-c @patrickvonplaten

### CODE REVIEW

File: src/transformers/modeling_bert.py [hunk 1]
```diff
@@ -223,6 +226,11 @@ def __init__(self, config):
 
         self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
 
+        self.position_embedding_config = config.position_embedding_config
+        if self.position_embedding_config > 0:
+            self.max_position_embeddings = config.max_position_embeddings
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:231]: no need to save `max_position_embeddings` here IMO -> you can use `self.config.max_position_embeddings` further down.

File: src/transformers/configuration_bert.py [hunk 2]
```diff
@@ -91,7 +91,11 @@ class BertConfig(PretrainedConfig):
             The epsilon used by the layer normalization layers.
         gradient_checkpointing (:obj:`bool`, `optional`, defaults to :obj:`False`):
             If True, use gradient checkpointing to save memory at the expense of slower backward pass.
-
+        position_embedding_config (:obj:`int`, `optional`, defaults to 0):
```
Comment (patrickvonplaten) [inline src/transformers/configuration_bert.py:94]: can we call it instead `position_embedding_type` with "absolute" (default), "relative", <good_name_for_method_4>?

**Comment by patrickvonplaten** [comment]
Hey @zhiheng-huang,

Thanks for the PR!

In general I'm fine with this PR - think adding more types of position embeddings to a model is OK. 
Also can you rebase your PR to the most current version of `master` - I think you are working on a rather old version.

File: src/transformers/modeling_bert.py (hunk 1 above)
Comment (zhiheng-huang) [reply src/transformers/modeling_bert.py:231]: max_position_embeddings is initialized here and it is used in line 273, where the self.config or config is not available in forward().

File: src/transformers/configuration_bert.py (hunk 2 above)
Comment (zhiheng-huang) [reply src/transformers/configuration_bert.py:94]: Good point, will do.

**Comment by zhiheng-huang** [comment]
> Hey @zhiheng-huang,
> 
> Thanks for the PR!
> 
> In general I'm fine with this PR - think adding more types of position embeddings to a model is OK.
> Also can you rebase your PR to the most current version of `master` - I think you are working on a rather old version.

It was rebased to master on 10/27, will rebase again for the new revision.

**Comment by zhiheng-huang** [comment]
@patrickvonplaten, I did a rebase and I am not sure this is the correct way to review commit "Address review comment". Please let me know if there is a better way to upload the new diff.

**Comment by patrickvonplaten** [comment]
Hey @zhiheng-huang,

I think there was a problem with the rebase it seems like added all commits on master on top of your PR. 
This happens from time to time sadly :-/ 
The way I'd fix it is to first save your changes of the last commit: https://github.com/huggingface/transformers/pull/8108/commits/ffe2e64c64f03c141cc085c8f3f509ae2e0992e2 somewhere (maybe a new branch).

Then correctly reset the head of your branch before all other commitns were falsely added:
```
git reset --hard 36729ee
```

Then add the single commit you saved in another branch


```
git cherry-pick ffe2e64
```

and finally either you correctly rebase OR the safer option here would probably be to merge the master into your branch

```
git fetch upstream/master
git merge upstream master
```

Hope this helps!

**Comment by zhiheng-huang** [comment]
> Hey @zhiheng-huang,
> 
> I think there was a problem with the rebase it seems like you added all commits on master on top of your PR.
> This happens from time to time sadly :-/
> 
> The way I'd fix it is to first save your changes of the last commit: [ffe2e64](https://github.com/huggingface/transformers/commit/ffe2e64c64f03c141cc085c8f3f509ae2e0992e2) somewhere (maybe a new branch).
> 
> Then correctly reset the head of your branch before all other commitns were falsely added:
> 
> ```
> git reset --hard 36729ee
> ```
> 
> Then add the single commit you saved in another branch
> 
> ```
> git cherry-pick ffe2e64
> ```
> 
> and finally either you correctly rebase OR the safer option here would probably be to merge the master into your branch
> 
> ```
> git fetch upstream/master
> git merge upstream master
> ```
> 
> Hope this helps!

Thanks. @patrickvonplaten. this helps but I may have to revert the commits merged to zhiheng-huang:transformers-relative-embedding. I created a new PR to keep cleaner history and we can move forward from there.

**Pull request closed without merging**